src/server/templates/zod.ts

```diff
@@ -0,0 +1,323 @@
+
+    return `z.${zodType}().nullable()`
+}
+
+function writeFunctions(
+    functions: PostgresFunction[],
+    types: PostgresType[],
+    arrayTypes: PostgresType[],
+): string {
+    const schemaFunctionsGroupedByName = functions.reduce((acc, curr) => {
+        acc[curr.name] ??= []
+        acc[curr.name].push(curr)
+        return acc
+    }, {} as Record<string, PostgresFunction[]>)
+
+    return `{
+        ${Object.entries(schemaFunctionsGroupedByName).map(([rawFnName, functions]) => {
+            const name = JSON.stringify(rawFnName)
+            
+            if (functions.length === 1) {
+                return `${name}: ${writeFunction(functions[0], types, arrayTypes)}`
+            }
+            
+            return `
+                ${name}: z.union([
+                    ${functions.map((func) => writeFunction(func, types, arrayTypes)).join(',\n')}
+                ])
+            `
+    }).join(',\n')}
+    }`
+}
+
+function writeFunction(func: PostgresFunction, types: PostgresType[], arrayTypes: PostgresType[]): string {
+    const inArgs = func.args.filter(({ mode }) => mode === 'in')
+
+    return `z.object({
+        ${inArgs.map(arg => `${JSON.stringify(arg.name)}: ${writeFunctionArg(arg, types, arrayTypes)}`).join(',\n')}
+    })`
+}
+
+function writeFunctionArg(arg: PostgresFunction['args'][0], types: PostgresType[], arrayTypes: PostgresType[]): string {
+    let type = arrayTypes.find(({ id }) => id === arg.type_id)
+    if (type) {
+        // If it's an array type, the name looks like `_int8`.
+        const elementTypeName = type.name.substring(1)
+        return `z.array(z.${basicZodType(elementTypeName)})` + (arg.has_default ? '.optional()' : '')
+    }
+    type = types.find(({ id }) => id === arg.type_id)
+    if (type) {
+        return "z." + basicZodType(type.format) + (arg.has_default ? '.optional()' : '')
+    }
+
+    console.debug(`Function: Unknown type ${arg.type_id}`)
+
+    return `z.unknown()` + (arg.has_default ? '.optional()' : '')
+}
+
+function basicZodType(pgType: string): string {
+    // Array
+    if (pgType.startsWith("_")) {
+        return basicZodType(pgType.substring(1)) + ".array()"
+    }
+
+    if ([ 'bool', 'boolean' ].includes(pgType)) {
+        return "boolean()"
+    }
+
+    if (['int2', 'int4', 'int8', 'float4', 'float8', 'numeric', 'integer', 'bigint', 'oid'].includes(pgType)) {
+        return 'number()'
+    }
+
+    if (
+        [
+            'bytea',
+            'bpchar',
+            'varchar',
+            'text',
+            'citext',
+            'uuid',
+            'vector',
+            'json',
+            'jsonb',
+            'inet',
+            'cidr',
+            'macaddr',
+            'macaddr8',
+            'character varying',
+        ].includes(pgType)
+    ) {
+        return 'string()'
+    }
+
+    if (["date", "time", "timetz", "timestamp", "timestamptz", "timestamp with time zone"].includes(pgType)) {
+        return 'date()'
+    }
```

Comment:
this is not correct. dates must be strings. for with timezone types we also need to set the `offset` parameter to true.

https://zod.dev/?id=iso-datetimes